[expression.dd] Improve IndexExpression & SliceExpression docs

ntrel opened the pull request:
* Make IndexExpression docs more generic.
* Mention pointer indexing & slicing.

* `$` is also declared for *ValueSeq* slices and can be for user types - make this paragraph independent of array slicing.
* Don't only talk about slicing arrays throughout, but slicing elements of the *PostfixExpression* instead. This is more generic.
* Mention first and last element expressions for arrays and pointers.
* Add *Slice Conversion to Static Array* subheading, move *ValueSeq* paragraph above it & add a *ValueSeq* link.

maxhaton commented on spec/expression.dd:
```diff
@@ -1266,43 +1266,51 @@ $(GNAME Slice):
+        the result of the slice is a new $(I ValueSeq) formed
+        from the upper and lower bounds, which must statically evaluate
+        to integral constants.
+        It is an error if those bounds are out of range.
```
```suggestion
        These bounds must be in range.
```

maxhaton replied:
Reads a little bit nicer + not a huge fan of language specs mentioning "error" and similar.

ntrel replied:
'must be' is not as clear that this is checked at compile time. (I didn't change the original sentence, just moved it).

dkorpel commented on spec/expression.dd:
```diff
@@ -1266,43 +1266,51 @@ $(GNAME Slice):
+
     $(P If the slice bounds can be known at compile time, the slice expression
-    is implicitly convertible to an lvalue of static array. For example:)
+    may be implicitly convertible to an lvalue of static array. For example:)
```
Why `may be`, are there cases where it isn't implicitly convertible?

ntrel replied:
@dkorpel The *PostfixExpression* might be a user-defined type where `opSlice` doesn't return a dynamic array.

ntrel commented:
This pull now covers `IndexExpression` as well as it's closely related. I also updated the first paragraph of the `SliceExpression` docs and added a paragraph about slicing a pointer.

dkorpel commented on spec/expression.dd:
```diff
@@ -1235,22 +1235,29 @@ $(GNAME IndexExpression):
+        dynamic array type, the result of the indexing is an lvalue
+        of the *i*th element in the array.
+        If $(I PostfixExpression) is a pointer `p`, the result is `*(p + i)`.
+        In each case, `i` is an integer evaluated from $(I ArgumentList).
```
You use `i` twice and only define what it refers to afterwards. I think that's a bit awkward, I'd rather see `i` defined first.

ntrel replied:
I have now defined `i` in the same sentence as the first use, hopefully that reads better.
